istio/istio
# use only isito chart to deploy all components

Use helm hooks to install CRD before installing the custom objects that depend on them. With this then there is not need to run `deployIstioHelmCharts.sh` to install istio, just `helm install istio` is enough.

TODO:

- [x]  Move all components into istio chart
- [x] Use post-install hook to install custom objects (that depend on CRD)
- [x] Move all needed manifests to `istio` chart
- [x] Full test

How to test it:
run `helm install istio` (on `install/kubernetes/helm` folder). 

helm hooks are not quite ready to handle CRD. Even if you define CRD as a `pre-install` hook, `helm isntall` will fail if it founds a custom object that depend on this CRD (so they use api "config.istio.io") in the same install. This is due to validation step that helm does before installing any object. Check kubernetes/helm#2925 (comment)

So, according to my knowledge, there two ways to install pass this issue using helm right now:

- A two install step:
First install CRD and then install the custom objects that depend on them. This is what the incubator's istio chart is doing with a `helm init` followed by a `helm upgrade` as you can see [here](https://github.com/kubernetes/charts/tree/master/incubator/istio#installing-the-chart).

- Install custom objects in a `post-install` hook:
In order to pass helm validation, you can create custom objects out of helm context by running a `post-intall` job that will manually create custom objects. This is what I choose in this PR, as you can see `create-custom-objects-job.yaml` job create custom object  by running `kubectl apply -f custom-objects` after helm installation. Also, this is what `prometheus-operator` is doing right now as you can see here in the [templates](https://github.com/coreos/prometheus-operator/tree/master/helm/prometheus-operator).

I'll open an issue in `helm` repo to see if there is a better way to do this. Let me know what you guys think and if you can install all by just running `helm install`.
